File: src/main/java/org/wysko/midis2jam2/instrument/guitar/Guitar.java
package org.wysko.midis2jam2.instrument.guitar;

import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;
import com.jme3.scene.Node;
import com.jme3.scene.Spatial;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.midi.MidiChannelSpecificEvent;

import java.util.HashMap;
import java.util.List;

import static org.wysko.midis2jam2.Midis2jam2.rad;

/**
 * The guitar. What more do you want?
 *
 * @see FrettedInstrument
 */
public class Guitar extends FrettedInstrument {
	
	static final Vector3f BASE_POSITION = new Vector3f(43.431f, 35.292f, 7.063f);
	
	public Guitar(Midis2jam2 context, List<MidiChannelSpecificEvent> events, GuitarType type) {
		super(context,
				new FrettingEngine(6, 22, new int[] {40, 45, 50, 55, 59, 64}, 40, 79),
				events,
				new FrettedInstrumentPositioning(16.6f,
						-18.1f,
						new Vector3f[] {
								new Vector3f(0.8f, 1, 0.8f),
								new Vector3f(0.75f, 1, 0.75f),
								new Vector3f(0.7f, 1, 0.7f),
								new Vector3f(0.77f, 1, 0.77f),
								new Vector3f(0.75f, 1, 0.75f),
								new Vector3f(0.7f, 1, 0.7f),
						},
						new float[] {-0.93f, -0.56f, -0.21f, 0.21f, 0.56f, 0.90f},
						new float[] {-1.55f, -0.92f, -0.35f, 0.25f, 0.82f, 1.45f},
						new FretHeightByTable(new HashMap<Integer, Float>() {{
							put(0, 0f);
							put(1, 0.03744493392f);
							put(2, 0.09691629956f);
							put(3, 0.1431718062f);
							put(4, 0.1916299559f);
							put(5, 0.2400881057f);
							put(6, 0.2841409692f);
							put(7, 0.3193832599f);
							put(8, 0.359030837f);
							put(9, 0.3920704846f);
							put(10, 0.4295154185f);
							put(11, 0.4603524229f);
							put(12, 0.4911894273f);
							put(13, 0.5176211454f);
							put(14, 0.5440528634f);
							put(15, 0.5726872247f);
							put(16, 0.5947136564f);
							put(17, 0.6189427313f);
							put(18, 0.6387665198f);
							put(19, 0.6585903084f);
							put(20, 0.6806167401f);
							put(21, 0.6982378855f);
							put(22, 0.7158590308f);
						}})),
				6,
				context.loadModel(type.modelFileName, type.textureFileName, Midis2jam2.MatType.UNSHADED, 0.9f)
		);
		
		
		for (int i = 0; i < 6; i++) {
			Spatial string;
			if (i < 3) {
				string = context.loadModel("GuitarStringLow.obj", type.textureFileName, Midis2jam2.MatType.UNSHADED, 0.9f);
			} else {
				string = context.loadModel("GuitarStringHigh.obj", type.textureFileName, Midis2jam2.MatType.UNSHADED, 0.9f);
			}
			upperStrings[i] = string;
			instrumentNode.attachChild(upperStrings[i]);
		}
		
		// Position each string
		final float forward = 0.125f;
		upperStrings[0].setLocalTranslation(positioning.topX[0], positioning.topY, forward);
		upperStrings[0].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(-1)));
		upperStrings[0].setLocalScale(positioning.restingStrings[0]);
		
		upperStrings[1].setLocalTranslation(positioning.topX[1], positioning.topY, forward);
		upperStrings[1].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(-0.62)));
		upperStrings[1].setLocalScale(positioning.restingStrings[1]);
		
		upperStrings[2].setLocalTranslation(positioning.topX[2], positioning.topY, forward);
		upperStrings[2].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(-0.22)));
		upperStrings[2].setLocalScale(positioning.restingStrings[2]);
		
		upperStrings[3].setLocalTranslation(positioning.topX[3], positioning.topY, forward);
		upperStrings[3].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(0.08)));
		upperStrings[3].setLocalScale(positioning.restingStrings[3]);
		
		upperStrings[4].setLocalTranslation(positioning.topX[4], positioning.topY, forward);
		upperStrings[4].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(0.45)));
		upperStrings[4].setLocalScale(positioning.restingStrings[4]);
		
		upperStrings[5].setLocalTranslation(positioning.topX[5], positioning.topY, forward);
		upperStrings[5].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(0.9)));
		upperStrings[5].setLocalScale(positioning.restingStrings[5]);
		
		// Lower strings
		for (int i = 0; i < 6; i++) {
			for (int j = 0; j < 5; j++) {
				if (i < 3)
					lowerStrings[i][j] = context.loadModel("GuitarLowStringBottom" + j + ".obj", type.textureFileName, Midis2jam2.MatType.UNSHADED, 0.9f);
				else
					lowerStrings[i][j] = context.loadModel("GuitarHighStringBottom" + j + ".obj", type.textureFileName, Midis2jam2.MatType.UNSHADED, 0.9f);
				instrumentNode.attachChild(lowerStrings[i][j]);
			}
		}
		
		// Position lower strings
		for (int i = 0; i < 5; i++) {
			lowerStrings[0][i].setLocalTranslation(positioning.bottomX[0], positioning.bottomY, forward);
			lowerStrings[0][i].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(-1)));
			lowerStrings[0][i].setLocalScale(positioning.restingStrings[0]);
		}
		for (int i = 0; i < 5; i++) {
			lowerStrings[1][i].setLocalTranslation(positioning.bottomX[1], positioning.bottomY, forward);
			lowerStrings[1][i].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(-0.62)));
			lowerStrings[1][i].setLocalScale(positioning.restingStrings[0]);
		}
		for (int i = 0; i < 5; i++) {
			lowerStrings[2][i].setLocalTranslation(positioning.bottomX[2], positioning.bottomY, forward);
			lowerStrings[2][i].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(-0.22)));
			lowerStrings[2][i].setLocalScale(positioning.restingStrings[0]);
		}
		
		for (int i = 0; i < 5; i++) {
			lowerStrings[3][i].setLocalTranslation(positioning.bottomX[3], positioning.bottomY, forward);
			lowerStrings[3][i].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(0.08)));
			lowerStrings[3][i].setLocalScale(positioning.restingStrings[0]);
		}
		
		for (int i = 0; i < 5; i++) {
			lowerStrings[4][i].setLocalTranslation(positioning.bottomX[4], positioning.bottomY, forward);
			lowerStrings[4][i].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(0.45)));
			lowerStrings[4][i].setLocalScale(positioning.restingStrings[0]);
		}
		
		for (int i = 0; i < 5; i++) {
			lowerStrings[5][i].setLocalTranslation(positioning.bottomX[5], positioning.bottomY, forward);
			lowerStrings[5][i].setLocalRotation(new Quaternion().fromAngles(0, 0, rad(0.9)));
			lowerStrings[5][i].setLocalScale(positioning.restingStrings[0]);
		}
		
		// Hide all wobbly strings
		for (int i = 0; i < 6; i++) {
			for (int j = 0; j < 5; j++) {
				lowerStrings[i][j].setCullHint(Spatial.CullHint.Always);
			}
		}
		
		
		// Initialize note fingers
		for (int i = 0; i < 6; i++) {
			noteFingers[i] = context.loadModel("GuitarNoteFinger.obj", type.textureFileName, Midis2jam2.MatType.UNSHADED, 0.9f);
			instrumentNode.attachChild(noteFingers[i]);
			noteFingers[i].setCullHint(Spatial.CullHint.Always);
		}
		
		// Position guitar
		highestLevel = new Node();
		highestLevel.setLocalTranslation(BASE_POSITION);
		highestLevel.setLocalRotation(new Quaternion().fromAngles(rad(2.66), rad(-44.8), rad(-60.3)));
		highestLevel.attachChild(instrumentNode);
		context.getRootNode().attachChild(highestLevel);
		
	}
	
	@Override
	public void tick(double time, float delta) {
		setIdleVisibilityByPeriods(finalNotePeriods, time, highestLevel);
		
		final int indexThis = getIndexOfThis();
		Vector3f add = new Vector3f(BASE_POSITION).add(new Vector3f(indexThis * 5, indexThis * -4, 0));
		highestLevel.setLocalTranslation(add);
		
		handleStrings(time, delta);
	}
	
	public enum GuitarType {
		ACOUSTIC("Guitar.obj", "GuitarSkin.bmp"), // TODO Obviously.
		ELECTRIC("Guitar.obj", "GuitarSkin.bmp");
		final String modelFileName;
		final String textureFileName;
		
		GuitarType(String modelFileName, String textureFileName) {
			this.modelFileName = modelFileName;
			this.textureFileName = textureFileName;
		}
	}
}

File: app/src/commonMain/kotlin/org/wysko/midis2jam2/instrument/family/guitar/Guitar.kt
/*
 * Copyright (C) 2025 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.guitar

import com.jme3.math.Quaternion
import com.jme3.math.Vector3f
import com.jme3.scene.Geometry
import com.jme3.scene.Spatial
import com.jme3.scene.Spatial.CullHint.Always
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import org.wysko.kmidi.midi.TimedArc
import org.wysko.kmidi.midi.event.MidiEvent
import org.wysko.kmidi.midi.event.NoteEvent
import org.wysko.midis2jam2.Midis2jam2
import org.wysko.midis2jam2.util.*
import org.wysko.midis2jam2.util.Utils.rad
import org.wysko.midis2jam2.util.resourceToString
import org.wysko.midis2jam2.world.STRING_GLOW
import org.wysko.midis2jam2.world.modelD
import kotlin.time.Duration

private val BASE_POSITION = v3(47, 35.3, 7.0)
private const val GUITAR_VECTOR_THRESHOLD = 8

private val GUITAR_MODEL_PROPERTIES: StringAlignment =
    Json.decodeFromString(resourceToString("/instrument/alignment/Guitar.json"))
private val GUITAR_CHORD_DEFINITIONS_STANDARD_E: Set<ChordDefinition> =
    Json.decodeFromString(resourceToString("/instrument/chords/Guitar.json"))
private val GUITAR_CHORD_DEFINITIONS_DROP_D: Set<ChordDefinition> =
    Json.decodeFromString<Set<ChordDefinition>>(resourceToString("/instrument/chords/Guitar.json"))
        .map { (notes, frets) ->
            ChordDefinition(
                notes = (listOf(if (notes[0].toInt() != -1) notes[0] - 2 else -1) + notes.subList(1, 6))
                    .map { it.toByte() },
                frets = frets
            )
        }.toSet()

/**
 * The Guitar.
 *
 * @param context The context to the main class.
 * @param events The list of all events that this instrument should be aware of.
 * @param type The type of guitar.
 * @see FrettedInstrument
 */
class Guitar(context: Midis2jam2, events: List<MidiEvent>, type: GuitarType) : FrettedInstrument(
    context = context,
    events = events,
    frettingEngine = StandardFrettingEngine(
        numberOfStrings = 6,
        numberOfFrets = 22,
        openStringMidiNotes = if (needsDropTuning(events)) GuitarTuning.DROP_D.values else GuitarTuning.STANDARD.values
    ),
    positioning = with(GUITAR_MODEL_PROPERTIES) {
        FrettedInstrumentPositioning(
            upperY = upperVerticalOffset,
            lowerY = lowerVerticalOffset,
            restingStrings = scales.map { Vector3f(it, 1f, it) }.toTypedArray(),
            upperX = upperHorizontalOffsets,
            lowerX = lowerHorizontalOffsets,
            fretHeights = FretHeightByTable.fromJson("Guitar")
        )
    },
    numberOfStrings = 6,
    instrumentBody = context.modelD(
        if (needsDropTuning(events)) type.modelDropD else type.model,
        type.texture
    ) to "GuitarSkin.bmp"
) {

    private val dictionary =
        if (needsDropTuning(events)) GUITAR_CHORD_DEFINITIONS_DROP_D else GUITAR_CHORD_DEFINITIONS_STANDARD_E
    private val openStringValues =
        if (needsDropTuning(events)) GuitarTuning.DROP_D.values else GuitarTuning.STANDARD.values

    /**
     * Maps each [TimedArc] that this Guitar is responsible to play to its [FretboardPosition].
     */
    override val notePeriodFretboardPosition: Map<TimedArc, FretboardPosition> =
        BetterFretting(context, dictionary, openStringValues, events).calculate(timedArcs)

    override val upperStrings: Array<Spatial> = Array(6) {
        context.modelD(if (it < 3) "GuitarStringLow.obj" else "GuitarStringHigh.obj", type.texture)
    }.apply {
        forEachIndexed { index, string ->
            geometry += string
            with(GUITAR_MODEL_PROPERTIES) {
                string.loc = v3(upperHorizontalOffsets[index], upperVerticalOffset, FORWARD_OFFSET)
                string.rot = v3(0, 0, -rotations[index])
            }
        }
    }

    override val lowerStrings: List<List<Spatial>> = List(6) { string ->
        List(5) { animFrame: Int ->
            context.modelD(
                if (string < 3) "GuitarLowStringBottom$animFrame.obj" else "GuitarHighStringBottom$animFrame.obj",
                type.texture
            ).also {
                geometry.attachChild(it)
                it.cullHint = Always
                (it as Geometry).material.setColor("GlowColor", STRING_GLOW)
            }
        }
    }.apply {
        indices.forEach { i ->
            repeat(5) { j ->
                with(this[i][j]) {
                    GUITAR_MODEL_PROPERTIES.let {
                        loc = v3(it.lowerHorizontalOffsets[i], it.lowerVerticalOffset, FORWARD_OFFSET)
                        rot = v3(0, 0, -it.rotations[i])
                    }
                }
            }
        }
    }

    override fun adjustForMultipleInstances(delta: Duration) {
        (updateInstrumentIndex(delta) * 1.5f).let {
            with(v3(5, -2, 0).mult(it)) {
                root.loc = when {
                    it < GUITAR_VECTOR_THRESHOLD -> this
                    else -> this.apply { setY(-2f * GUITAR_VECTOR_THRESHOLD) }
                }
            }
        }
    }

    /**
     * The type of guitar.
     */
    sealed class GuitarType(
        internal val model: String,
        internal val modelDropD: String,
        internal val texture: String
    ) {
        /** Acoustic guitar type. */
        data object Acoustic : GuitarType("GuitarAcoustic.obj", "GuitarAcousticDropD.obj", "AcousticGuitar.png")

        /** Clean guitar type. */
        data object Clean : GuitarType("Guitar.obj", "GuitarD.obj", GuitarSkin["clean"].file)

        /** Jazz guitar type. */
        data object Jazz : GuitarType("Guitar.obj", "GuitarD.obj", GuitarSkin["jazz"].file)

        /** Muted guitar type. */
        data object Muted : GuitarType("Guitar.obj", "GuitarD.obj", GuitarSkin["muted"].file)

        /** Overdrive guitar type. */
        data object Overdriven : GuitarType("Guitar.obj", "GuitarD.obj", GuitarSkin["overdriven"].file)

        /** Distortion guitar type. */
        data object Distortion : GuitarType("Guitar.obj", "GuitarD.obj", GuitarSkin["distortion"].file)

        /** Harmonics guitar type. */
        data object Harmonics : GuitarType("Guitar.obj", "GuitarD.obj", GuitarSkin["harmonics"].file)
    }

    init {
        /* Position guitar */
        geometry.localTranslation = BASE_POSITION
        geometry.localRotation = Quaternion().fromAngles(rad(2.66), rad(-44.8), rad(-60.3))
    }
}

private fun needsDropTuning(events: List<MidiEvent>): Boolean =
    (events.filterIsInstance<NoteEvent.NoteOn>().minByOrNull { it.note }?.note ?: 127) < 40

private enum class GuitarTuning(val values: IntArray) {
    STANDARD(intArrayOf(40, 45, 50, 55, 59, 64)),
    DROP_D(intArrayOf(38, 45, 50, 55, 59, 64))
}
